src/Ui/UiRequest.py

```diff
@@ -285,24 +299,41 @@ def sendHeader(self, status=200, content_type="text/html", noscript=False, allow
             headers["Access-Control-Allow-Headers"] = "Origin, X-Requested-With, Content-Type, Accept, Cookie, Range"
             headers["Access-Control-Allow-Credentials"] = "true"
 
-        if content_type == "text/html":
-            content_type = "text/html; charset=utf-8"
-        if content_type == "text/plain":
-            content_type = "text/plain; charset=utf-8"
-
         # Download instead of display file types that can be dangerous
         if re.findall("/svg|/xml|/x-shockwave-flash|/pdf", content_type):
             headers["Content-Disposition"] = "attachment"
-
-        cacheable_type = (
-            content_type == "text/css" or content_type.startswith("image") or content_type.startswith("video") or
-            self.env["REQUEST_METHOD"] == "OPTIONS" or content_type == "application/javascript"
+            
+        # UTF-8 character encoding
+        if content_type in ("text/plain", "text/html", "text/css", "application/javascript", "application/json", "application/manifest+json"):
+            content_type = "%s;" % content_type + " " + "charset=utf-8"
+            
+        # Caching
+        cache_txt_css_html = (
+            content_type.startswith("text")
+        )
+        cache_image = (
+            content_type.startswith("image")
+        )
+        cache_font = (
+            content_type.startswith("font")
+        )
+        cache_video = (
+            content_type.startswith("video")
+        )
+        nocache_application = (
+            content_type.startswith("application")
         )
 
-        if status in (200, 206) and cacheable_type:  # Cache Css, Js, Image files for 10min
-            headers["Cache-Control"] = "public, max-age=600"  # Cache 10 min
-        else:
-            headers["Cache-Control"] = "no-cache, no-store, private, must-revalidate, max-age=0"  # No caching at all
+        if status in (200, 206) and cache_txt_css_html:  # Cache MIME type text/* for 1 week
```

Comment:
AFAIK we don't do cache invalidation so this might be a problem. Static site update might take a long while...

Reply:
Cache is `public` as you see. In the other hand after `siteSign` there should be `sitePublish` which will notify all peers about the changes.

Reply:
How does this help? I mean, the browser would still use cached html, wouldn't it?

Reply:
Yes, but not for 10 minutes. It would be crazy to assume that a static sites changes in every 10 minutes. 
Does the browser cache `content.json`? No, is not because we are not caching any MIME type **application/***, specially not JavaScript.

Reply:
It's crazy to assume 10 minutes but it's crazy to assume a whole week too. I'd recommend 2 hours or so.

Reply:
How is content.json related? It's not accessed by browser at all.

Reply:
Ok, it looks like you thought of site updating as of publishing. I actually meant that users would receive updates but their browsers wouldn't show the changes.

Reply:
Note:
On line 299 in `cacheable_type` JavaScript was `application/javascript`
`            self.env["REQUEST_METHOD"] == "OPTIONS" or content_type == "application/javascript"`

On line 272 which was also [removed](https://github.com/HelloZeroNet/ZeroNet/pull/2248/commits/091f30fbd7163a28736e970f9a432b67d4dab4c7), JavaScript used as `text/javascript`
`        if content_type == "text/javascript" and not self.env.get("HTTP_REFERER"):`

Reply:
> Ok, it looks like you thought of site updating as of publishing. I actually meant that users would receive updates but their browsers wouldn't show the changes.

Depend, I seen that many people rocking with Tor Browser. :man_technologist: 
Caching is handy when you have a gateway or ZeroNet is behind a reverse proxy.

Reply:
What about making a `--cache_for` configuration option then? @HelloZeroNet

Reply:
I can add `stale-while-revalidate` to MIME type **text/*** if that helps.

Reply:
Ok, I guess that makes sense. I'd still add `--cache_for` though, but I guess it should be another PR.

Reply:
@imachug I don't think that need `--cache_for`. 
I do have 200+ static websites and I can tell you for sure that I'm not updating them every 10 minutes. Not even in one day. Caching is good how you see in these changes.

Reply:
In case of ZeroNet, static websites are often actually made with jekyll or similar tools (look at @caryoscelus'es blog). *These* kinds of sites are updated rather often.

Reply:
Nothing is stopping him in lowering the default `max-age` value if needed.

Reply:
Who is "him"? @caryoscelus?

Reply:
Yes. You come up with one specific example and I replied accordingly. 

According to my checks by refreshing the site in the browser you will eventually get a fresh copy of the static resources. 
When you use _back_ and _forward_ button than obviously you will see the cached version. 

See in ZeroNet `frame` and using `raw`.

Reply:
> Yes. You come up with one specific example and I replied accordingly.

caryoscelus (or whoever runs a site) can't just change caching settings on *others'* computers. How would you even do that?

> According to my checks by refreshing the site in the browser you will eventually get a fresh copy of the static resources.

This "eventually" often includes accidentally pressing Shift / Ctrl or doing something in web console. That's not something that an average user should think about.

> See in ZeroNet frame and using raw.

Now what?.. Are you asking me to use raw mode? That's often not an option -- static sites *can* include JS, e.g. for animations.

Reply:
Let's say you have a chat in ZeroNet and you send `text/html` to the cache for one week while using JavaScript to print messages. Not to mention you can use JavaScript to refresh the site time-to-time without needing to mess around the default caching of the resources. Correct me if I'm wrong.

Reply:
Let's talk about a specific example -- a static blog. Pretend that I had this blog on clearnet (e.g. on GitHub pages) but then I moved it to ZeroNet. So, my blog doesn't work with ZeroNet API, it's just a collection of html files and a few JS files that are used for animations and other interactive content. Now, how would you reload site content when a new post is publushed?

Reply:
No MIME type starts with `application/*` is cached! This includes JavaScript, JSON an others like `application/octet-stream`. 
You can use query string on static resources like ZeroNet does, example: **all.css?rev=666**.

Reply:
We're talking about *HTML pages*, not scripts!

Reply:
Let me be very clear on that ZeroNet is not in development in order to provide hosting for Jekyll! If you want a blog than there are a tons of options. You can **clone ZeroTalk** for example or just go and run it on WordPress. 
Most people don't even use this frameworks and it will be fatal to assume that HTML pages must work like a dynamic blog! I say it again. The proposed caching is good.

Reply:
What the heck? ZeroNet aims at easy single-click zite creation. It should be easy to both create sites from scratch and copy clearnet sites. You can't just say "we won't do that"!

Reply:
I can also tell you this in Hungarian if you need. I born in Hungary, but left the country when Orban come into power nearly 10 years ago. My sister is a "border hunter" in the Croatian border, hunting real living people. I strongly believe that ZeroNet can really hurt those in power and I will not going to argue with you about how long the HTML should be cached. 

I, for that matter will always use this caching because as I said I do have 200+ static websites. It is already a pain to sign the new content and I don't want to see my server overloaded with request for static resources like HTML, CSS, video and other which are currently fetched every 10 minutes.

Reply:
Ok, that's why I asked to add a `--cache_for` option. It could be set to 2 hours by default and increased by those who have lots of sites.

Reply:
> Most people don't even use this frameworks and it will be fatal to assume that HTML pages must work like a dynamic blog! I say it again. The proposed caching is good.

Yeah, you're an idiot.

Reply:
Choose your words you retard!

Reply:
I will send another PR but the caching stays how it is. Nobody and I say it again NOBODY in this word changes HTML pages every 10 minutes or so. You can use JavaScript or whatever you want if you need to update the text. Show me an example HTML site where the text changes every 10 minutes otherwise be silent.

Reply:
I'm not the one here who thinks that everything *has* to use javascript to load content.

And you know what ZeroNet **is** in development to do... provide websites to users! LOL Jekyll is a generator for a website, right?

Finally, Jekyll isn't what he was talking about - Jekyll loads things dynamically via javascript I believe (I might be wrong about this).

Reply:
I will send another PR but the caching stays how it is. Nobody and I say it again NOBODY in this world changes HTML pages every 10 minutes or so. You can use JavaScript or whatever you want if you need to update the text. Show me an example HTML site where the text changes every 10 minutes otherwise be silent.

Reply:
Go fckyourself with you Jekyll.

Reply:
If you think that static zites aren't updated more than every week, then yeah... you're an idiot. If you think that zites have control over caching of stuff, then you're just assuming ZeroNet has this features without ever seeing it anywhere.

Reply:
You know what I will not send a PR because you are a disgusting idiot who probably never in his life wrote a website from 0 with HTML. You just talk shit about programs that generate HTML sites but you fail to come up with an example where HTML is changes every 10 minutes. 

ZeroNet should check the content.json and if the hash is different than refresh the site. That would be the solution but instead of writing the code up you are here to spit shit.

Reply:
> ZeroNet should check the content.json and if the hash is different than refresh the site. 

You clearly have no idea what you're talking about. LOL

Reply:
> sites but you fail to come up with an example where HTML is changes every 10 minutes.

You know how I can tell you're a **very fishy** person... because you keep spewing that everyone's saying "10 minutes" when nobody has said that. Everyone here has only said changes happen shorter than **a week**.

I'll be making my own PR.

Btw, I **did** have a static zite, lol. And it changed frequently.

Reply:
@krixano Seriously, you opened a fucking issue about font loading and I did fixed it. You are an idiot and you know it because you was unable to fix such a simple thing as font loading in Tor Browser.

Reply:
You make your own PR but copy nothing from mine!

Reply:
Right... Because apparently everyone who opens an issue also can't program? Because apparently people you think that there's **no other** reason for not fixing the issue themselves. Because apparently people don't have work or school or anything else going on in their life. Again a **very fishy person**.

Btw, there's a different between fixing something and half-assing it (and deliberately preventing things from work because you're an asshole).

Reply:
I think you should be kicked out of this repo because you appreciate nothing but talking shit. ZeroNet should really distance itself from people like you. You ruining it not developing. What developing, you don't know even how to fix font loading or where to look. You are a disaster and you look like one according to your photo, website etc..

Reply:
Right... says the person who also said crap about ivanq who is the second highest for most commits and has fixed more vulnerabilities than you've even found in ZeroNet, LOL

Reply:
I'm sorry? I come to this repo 2 days ago! 
Sorry that there is no a million PR already. How can be someone so retarded like you?

Reply:
> You make your own PR but copy nothing from mine!

Does this mean you're claiming copyright for your commit?

@HelloZeroNet I'd be weary of the potential legal consequences of this in the future. This is precisely why open-source repos have a file that requires all committers to give up rights to their committs.

Reply:
I'm Hungarian too just that you know.